### ISSUE TITLE
Cannot use React development mode

### ISSUE THREAD

**Issue by AlienRecall** [wailsapp/wails#999]
**Description**
Hi, I was trying to start my application with `wails dev` and using development mode on React, when I run `wails dev`, it executes the `nom run start` command that I wrote in `wails.json` but after that the application doesn't start

**To Reproduce**
Steps to reproduce the behaviour (If you use React in your wails project):
1. Write `npm run start` as "frontend:dev" command in wails.json
2. Set the same port as wails in your `start` command on package.json
3. Run `wails dev`

**Expected behaviour**
The app executes normally, with wails watcher for backend changes and react watcher for frontend changes

**System Details**
OS:             Windows 10 Home
Version:        2009 (Build: 22000)
ID:             21H2
Go Version:     go1.17.3
Platform:       windows
Architecture:   amd64

- [ ] This issue is for Windows and I have read the [Windows Developer Guide](https://wails.app/guides/windows/)

**Comment by leaanthony** [comment]
Can you identify how this is a Wails bug? Thanks.

**Comment by AlienRecall** [comment]
> Can you identify how this is a Wails bug? Thanks.

Basically when I run wails dev I see that wails executes the npm run start command and then it stops, I think that happens because npm run start remains active for frontend changes, is not possible to use concurrency to run npm run start and then wails dev?

I'm sorry if this doesn't seems like a bug but I didn't know where I have to notify things like this

**Comment by AlienRecall** [comment]
Or I use wails since the 1.0 and I remember that when I was developing I execute the dev command in one terminal and nom run start in another one terminal, maybe it could be a way also in this version?

**Comment by leaanthony** [comment]
Run in another terminal for now. What do you think a good solution would be? We could add a script that's run in parallel, but what is it called, when does it run and when should it be restarted if at all?

**Comment by Code-Ac** [comment]
I think the problem is that the react dev command is executed in the same thread as wails - so it just runs and runs whitout exiting waiting for changes. Meanwhile wails is waiting for the command to exit to continue whit the startup routine. I would recommend adding a `&` on the end of the `npm run dev` so it disconnects to a new task and lets wails continue whit startup. I don't really know how this would affect the shutdown routine because it's now a different task now.

**Comment by leaanthony** [comment]
There's been a ton of requests and talk around running a dev server. If someone wants to take on that piece, define what it should do then I will gladly support it. The original request was to have a dev specific build command and that's what we currently have. Many people have requested that this could be used to spawn an external build server. As you know, this is difficult because the build process blocks on this command, as was the original intent. So all this needs is a little thought and someone to want it bad enough to define what it should do and how that might happen in a backwards compatible way.

**Comment by Code-Ac** [comment]
Well, I don't really know how wails is working so based off what I saw until now my wishes would be:
1) To have a dev-server command starting a separate environment (although this can already be done by adding a `&` character on the end of the normal dev command so I don't see that much need for it)
2) Secondly and more importantly adding a wait-on option to wait for the in `devserverurl` defined server to start and deliver usable content (especially bigger projects take a bit of time to start up their development server)
3) A way to print out the outputs of the dev command if you would decide to run it in a different thread by default
4) Remove the compiling and starting of an internal development server when the `devserverurl` property is passed, it should not be needed anymore then.

**Comment by flin7** [comment]
Use webpack in watch mode with babel-loader instead of CRA.

**Comment by leaanthony** [comment]
> Use webpack in watch mode with babel-loader instead of CRA.

Is this how you set up your Wails project? Would you be interested in writing a guide?

**Comment by leaanthony** [comment]
@Code-Ac 

> 1. To have a dev-server command starting a separate environment (although this can already be done by adding a `&` character on the end of the normal dev command so I don't see that much need for it)

Does shutdown actually kill that process correctly if you use `&`? Also, this isn't a cross-platform way to do this. I'm wondering if adding a "RunDevCommandInSeparateThread" option might be a good way to indicate this.

> 2. Secondly and more importantly adding a wait-on option to wait for the in `devserverurl` defined server to start and deliver usable content (especially bigger projects take a bit of time to start up their development server)

Ah that's a nice idea. So it would just poll the port until it got a response?

> 3. A way to print out the outputs of the dev command if you would decide to run it in a different thread by default (see 1.)

I think this should be easy enough.

> 4. Remove the compiling ~and starting of an internal development server~ when the `devserverurl` property is passed, it should not be needed anymore then.

Agreed.

> 5. Maybe add an option to not start watchers on the frontent (not reload when saved) because from my knowledge most modern web development tools are doing that for themselves.

That's easy enough. I think that should be default if `devserverurl` is set.

One Question: If you are developing this way, how are you getting the bindings and runtime injected into your code that's served from the external dev server? That's been the tricky part so far. Options have been to either add `script` tags that point to the local dev server (won't work any more if disabled by default) or using the special bundled files in `wailsjs` dir (I've had no reports of anyone actually doing this).

**Comment by flin7** [comment]
> > Use webpack in watch mode with babel-loader instead of CRA.
> 
> @flin7 - Is this how you set up your Wails project? Would you be interested in writing a guide?

@leaanthony I've created a template for this:
https://github.com/flin7/wails-react-template

**Comment by Code-Ac** [comment]
> Does shutdown actually kill that process correctly if you use `&`? Also, this isn't a cross-platform way to do this. I'm wondering if adding a "RunDevCommandInSeparateThread" option might be a good way to indicate this.

I don't think it does, but it could be a workaround as long as it works for @flin7. Otherwise yeah a 

> Ah that's a nice idea. So it would just poll the port until it got a response?

Right. Maybe there is a better way of doing that by timeouts but idk. 

> I think this should be easy enough.

Maybe just pass them like the default dev server does.

> That's easy enough. I think that should be default if `devserverurl` is set.

Yeah, I am currently experiencing a few problems whit that (#1036)

> If you are developing this way, how are you getting the bindings and runtime injected into your code that's served from the external dev server?

Yeah, I don't know that much about the application yet. Thought that it would use a socket to communicate (->Reconnect; done) but it sounds like you are injecting the bindings and runtime somehow. As the app would reload on go-side changes anyways it would just get loaded again? (I literally don't know anything about the in-depth workings of this app yet. I could be and think I am completely wrong here)

**Comment by leaanthony** [comment]
> > > Use webpack in watch mode with babel-loader instead of CRA.
> > 
> > 
> > @flin7 - Is this how you set up your Wails project? Would you be interested in writing a guide?
> 
> @leaanthony I've created a template for this: https://github.com/flin7/wails-react-template

@AlienRecall - does this help?
